### PR TITLE
refactor proof tree formatting

### PR DESCRIPTION
**Pull request by lcnr**
mostly:
- handle indentation via a separate formatter
- change nested to use a closure

tested it after rebasing on top of #113536 and everything looks good.

r? @BoxyUwU

### CODE REVIEW

File: compiler/rustc_middle/src/traits/solve/inspect/format.rs [hunk 1]
```diff
@@ -1,11 +1,19 @@
 use super::*;
 
 pub(super) struct ProofTreeFormatter<'a, 'b> {
-    pub(super) f: &'a mut (dyn Write + 'b),
-    pub(super) on_newline: bool,
+    f: &'a mut (dyn Write + 'b),
 }
 
-impl Write for ProofTreeFormatter<'_, '_> {
+/// A formatter which adds 4 spaces of identation to its input before
```
Comment (lqd) [inline compiler/rustc_middle/src/traits/solve/inspect/format.rs:7]: tiny typo. 

```suggestion
/// A formatter which adds 4 spaces of indentation to its input before
```

**Comment by lcnr** [comment]
this doesn't make sense to me :sweat_smile: testing clippy passes for me xx

**Comment by BoxyUwU** [comment]
r=me when clippy decides to stop bricking rust-lang/rust's entire bors queue

**Pull request merged**